### PR TITLE
HttpBackgroundJob helper component

### PR DESCRIPTION
**Pull request by jtkech**
For now this component is not used at all in our source code.

It allows to execute a background job in an isolated ShellScope after the current HTTP request is completed. It allows a controller (or razor page) action to return without waiting for a job to complete, and then, if the job update some state data, also allows to render the job state.

And just a copy paste from `ModularBackgroundService` to make 2 or 3 helper extensions to not have duplicate code.

### CODE REVIEW

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+using Microsoft.AspNetCore.Http;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Builders;
+using OrchardCore.Modules;
+
+namespace OrchardCore.BackgroundTasks;
+
+internal static class ShellContextExtensions
+{
+    public static HttpContext CreateHttpContext(this ShellContext shell)
```
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs:12]:
```suggestion
    private const string Localhost = "localhost";

    private static readonly char[] _urlHostSeparators = new[] { ',', ' ' }

    public static HttpContext CreateHttpContext(this ShellContext shell)
```

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+        return context;
+    }
+
+    private static HttpContext CreateHttpContext(ShellSettings settings)
+    {
+        var context = new DefaultHttpContext().UseShellScopeServices();
+
+        context.Request.Scheme = "https";
+        var urlHost = settings.RequestUrlHost?.Split(new[] { ',', ' ' }, StringSplitOptions.RemoveEmptyEntries).FirstOrDefault();
```
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs:30]:
```suggestion
        var urlHost = settings.RequestUrlHost?.Split(_urlHostSeparators, StringSplitOptions.RemoveEmptyEntries).FirstOrDefault();
```

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        context.Request.Host = new HostString(urlHost ?? "localhost");
```
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs:31]:
```suggestion
        context.Request.Host = new HostString(urlHost ?? Localhost);
```

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/HttpContextExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+using System;
+using Microsoft.AspNetCore.Http;
+
+namespace OrchardCore.BackgroundTasks;
+
+internal static class HttpContextExtensions
+{
+    public static void SetBaseUrl(this HttpContext context, string baseUrl)
```
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/HttpContextExtensions.cs:8]: Can we check if the `baseUrl` is null or empty?

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundJobs/HttpBackgroundJob.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+using OrchardCore.Environment.Shell.Scope;
+
+namespace OrchardCore.BackgroundJobs;
+
+public static class HttpBackgroundJob
+{
+    /// <summary>
+    /// Executes a background job in an isolated <see cref="ShellScope"/> after the current HTTP request is completed.
+    /// </summary>
+    public static Task AfterEndOfRequestAsync(string jobName, Func<ShellScope, Task> job)
```
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Abstractions/BackgroundJobs/HttpBackgroundJob.cs:17]: `ExecuteAfterEndOfRequestAsync()` or `RunAfterEndOfRequestAsync`

**Review by hishamco — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundJobs/HttpBackgroundJob.cs (hunk 5 above)
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Abstractions/BackgroundJobs/HttpBackgroundJob.cs:17]: Hmm, just to not be too long, the usage being

    HttpBackgroundJob.AfterEndOfRequestAsync()

In place of

    HttpBackgroundJob.ExecuteAfterEndOfRequestAsync()

But why not, no closed opinion but for now waiting for other feedbacks

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundJobs/HttpBackgroundJob.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+        // Fire and forget in an isolated child scope.
+        _ = ShellScope.UsingChildScopeAsync(async scope =>
+        {
+            var timeoutTask = Task.Delay(60_000);
+
+            // Wait for the current 'HttpContext' to be released with a timeout of 60s.
+            while (httpContextAccessor.HttpContext != null)
+            {
+                await Task.Delay(1_000);
```
[the remark this answers is not in the document]
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Abstractions/BackgroundJobs/HttpBackgroundJob.cs:42]: What's the difference?

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/HttpContextExtensions.cs (hunk 4 above)
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/HttpContextExtensions.cs:8]: We use `Uri.TryCreate()`

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
+using System;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Builders;
+using OrchardCore.Modules;
+
+namespace OrchardCore.BackgroundTasks;
```
[the remark this answers is not in the document]
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs:8]: Hmm, no closed opinion but for now waiting for other feedbacks.

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs (hunk 1 above)
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs:16]: Okay done.

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs (hunk 2 above)
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs]: Okay done.

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs (hunk 3 above)
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs]: Okay done.

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundJobs/HttpBackgroundJob.cs (hunk 5 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Abstractions/BackgroundJobs/HttpBackgroundJob.cs:17]: We could rename the class to `HttpBackgroundJobRunner`, but again using verbs for methods is a best practice

Comment (jtkech) [reply src/OrchardCore/OrchardCore.Abstractions/BackgroundJobs/HttpBackgroundJob.cs:17]: Okay I will think about it and then change the name ;)

Comment (jtkech) [reply src/OrchardCore/OrchardCore.Abstractions/BackgroundJobs/HttpBackgroundJob.cs]: Okay done

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundJobs/HttpBackgroundJob.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Abstractions/BackgroundJobs/HttpBackgroundJob.cs:42]: Just a new line before the considtion ;)

**Review by hishamco — APPROVED** [review]
LGTM except using namespace block everywhere just for consistency, but not a big issue ;)

File: src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Skrypt) [reply src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs:8]: That's a topic for https://github.com/OrchardCMS/OrchardCore/pull/10724
For now, we should keep brackets for consistency.
Else, we can plan to do a code cleanup from VS with this new coding style and change them all if we decide so.

**Pull request merged**